File: client/android/app/src/main/java/app/mobiling/client/attachment/AttachmentFeatureBridge.kt
package app.mobiling.client.attachment

import app.mobiling.client.contract.attachment.AttachmentFileHandoffPayload
import app.mobiling.client.contract.attachment.AttachmentLinkPayload
import app.mobiling.client.contract.attachment.AttachmentLinkRequest
import app.mobiling.client.contract.attachment.AttachmentListPayload
import app.mobiling.client.contract.attachment.AttachmentUploadHandoffPayload
import app.mobiling.client.contract.attachment.AttachmentUploadHandoffRequest
import app.mobiling.client.data.attachment.AttachmentReader
import app.mobiling.client.data.attachment.AttachmentWriter
import app.mobiling.client.usecase.attachment.AttachmentAttachUseCase
import app.mobiling.client.usecase.attachment.AttachmentFileHandoffUseCase
import app.mobiling.client.usecase.attachment.AttachmentLoadListUseCase
import app.mobiling.client.usecase.attachment.AttachmentUploadHandoffUseCase

/**
 * Marketing America Corp. Oleksandr Tishchenko
 */
class AttachmentFeatureBridge(
    private val reader: AttachmentReader,
    private val writer: AttachmentWriter,
) {
    suspend fun list(
        ownerType: String,
        ownerId: String,
        context: String? = null,
        slot: String? = null,
    ): AttachmentListPayload =
        AttachmentLoadListUseCase(reader).invoke(ownerType, ownerId, context, slot)

    suspend fun attach(request: AttachmentLinkRequest): AttachmentLinkPayload =
        AttachmentAttachUseCase(writer).invoke(request)

    suspend fun fileHandoff(attachmentId: String): AttachmentFileHandoffPayload =
        AttachmentFileHandoffUseCase(writer).invoke(attachmentId)

    suspend fun uploadHandoff(request: AttachmentUploadHandoffRequest): AttachmentUploadHandoffPayload =
        AttachmentUploadHandoffUseCase(writer).invoke(request)
}

File: client/ios/App/Attachment/AttachmentFeatureBridge.swift
import Foundation

public struct AttachmentFeatureBridge: Sendable {
    private let reader: AttachmentReader
    private let writer: AttachmentWriter

    public init(reader: AttachmentReader, writer: AttachmentWriter) {
        self.reader = reader
        self.writer = writer
    }

    public func list(ownerType: String, ownerId: String, context: String? = nil, slot: String? = nil) async throws -> AttachmentListPayload {
        try await LoadAttachmentListUseCase(reader: reader)(ownerType: ownerType, ownerId: ownerId, context: context, slot: slot)
    }

    public func attach(request: AttachmentLinkRequest) async throws -> AttachmentLinkPayload {
        try await AttachAttachmentUseCase(writer: writer)(request: request)
    }

    public func fileHandoff(attachmentId: String) async throws -> AttachmentFileHandoffPayload {
        try await writer.fileHandoff(attachmentId: attachmentId)
    }

    public func uploadHandoff(request: AttachmentUploadHandoffRequest) async throws -> AttachmentUploadHandoffPayload {
        try await writer.uploadHandoff(request: request)
    }

    public func upload(request: AttachmentUploadHandoffRequest, fileName: String, mimeType: String, data: Data) async throws -> AttachmentItemPayload {
        try await writer.uploadAttachment(request: request, fileName: fileName, mimeType: mimeType, data: data)
    }
}
